README.md
# NAME

Nuvol - A cloud toolkit

# SYNOPSIS

    use Nuvol;

    # connect to a service
    my $configfile = '/path/to/configfile';
    my $service    = '...';  # one of Dropbox, Office365
    my $connector  = Nuvol::autoconnect($configfile, $service);

    # get main drive
    my $drive = $connector->drive('~');

    # upload a file
    use Mojo::File 'path';
    my $file = $drive->item('/My Text.txt')->copy_from(path 'Text on my PC.txt');

    # copy to another file
    my $file_2 = $file->copy_to('/path/to/Text Copy.txt');

    # download
    my $downloaded = $file_2->copy_to(path 'Downloaded Text.txt');

    # change the text
    $file_2->spurt('This text was changed.');

    # read it
    my $content = $file_2->slurp;

    # and delete the file
    $file_2->remove;

# DESCRIPTION

[Nuvol](https://metacpan.org/pod/Nuvol) is a toolkit to manipulate files and folders on cloud services. For
the beginning it supports [Dropbox](https://metacpan.org/pod/Nuvol%3A%3ADropbox), [Office
365](https://metacpan.org/pod/Nuvol%3A%3AOffice365), and a [Dummy service](https://metacpan.org/pod/Nuvol%3A%3ADummy).

    Nuvol
    └── Connector
        ├── Config
        │   └── config file
        └── Drive
            └── Item
                ├── File
                └── Folder

The services are organized in connectors, drives, items, files, and folders. The data needed to
access a service is stored in a config file.

- Connector

    The [Connector](https://metacpan.org/pod/Nuvol%3A%3AConnector) is responsible for authentication and for the connection to the
    cloud service.

- Config and config file

    The config file stores the tokens and other parameters used to establish a connection. Internally is
    is represented by a [Config](https://metacpan.org/pod/Nuvol%3A%3AConfig) object.

    **Warning:** The information in the config file allows full access to your cloud data for anyone who
    can read it. It should be stored at a secure place. Services that are no longer used should be
    disabled with ["disconnect" in Nuvol::Connector](https://metacpan.org/pod/Nuvol%3A%3AConnector#disconnect).

- Drive

    A [Drive](https://metacpan.org/pod/Nuvol%3A%3ADrive) is an isolated area where your data is stored. You may have a drive for your
    personal and another for your business data. Not all cloud providers support different drives.

- Item

    Every object in a drive is an [Item](https://metacpan.org/pod/Nuvol%3A%3AItem). Item is just an abstract type, a real object is
    either a [File](https://metacpan.org/pod/Nuvol%3A%3ARole%3A%3AFile) or a [Folder](https://metacpan.org/pod/Nuvol%3A%3ARole%3A%3AFolder).

    The syntax for drive items is oriented at [Mojo::File](https://metacpan.org/pod/Mojo%3A%3AFile), so anyone familiar with this module will
    recognize most of the methods.

# FUNCTIONS

None of the functions is exported.

## autoconnect

    use Nuvol;
    $connector = Nuvol::autoconnect($configfile, $service);

Opens a connection using an existing config file, or starts an interactive
authentication process if the file doesn't exist. Returns a
[Nuvol::Connector](https://metacpan.org/pod/Nuvol%3A%3AConnector).

## connect

    use Nuvol;
    $connector = Nuvol::connect($configfile);

Opens a connection using an existing config file. Returns a [Nuvol::Connector](https://metacpan.org/pod/Nuvol%3A%3AConnector).

# AUTHOR & COPYRIGHT

© 2013–2020 by Tekki (Rolf Stöckli).

This program is free software, you can redistribute it and/or modify it under the terms of the
Artistic License version 2.0.

# SEE ALSO

[Nuvol::Connector](https://metacpan.org/pod/Nuvol%3A%3AConnector), [Nuvol::Drive](https://metacpan.org/pod/Nuvol%3A%3ADrive), [Nuvol::Item](https://metacpan.org/pod/Nuvol%3A%3AItem), [Nuvol::Test](https://metacpan.org/pod/Nuvol%3A%3ATest).


Changes
Revision history for perl distribution Nuvol

0.03 Not Released
  - Nuvol::autoconnect function

0.02 2020-02-24T06:48:10+0100
  - Support for Dropbox
  - Method for realpath
  - Paths are absolute
  - Type flags renamed to is_file and is_folder

0.01 2020-02-12T17:56:13+0100
  - First Release with basic functionality for Office 365


lib/Nuvol/Test/Roles.pm
package Nuvol::Test::Roles;
use Mojo::Base -base, -signatures;

use Exporter 'import';
our @EXPORT      = ();
our @EXPORT_OK   = ();
our %EXPORT_TAGS = (all => \@EXPORT_OK);
for my $role (qw|file folder metadata|) {
  push @EXPORT_OK, map {"test_${role}_$_"} qw|applied methods prerequisites|;
  $EXPORT_TAGS{$role} = [map {"test_${role}_$_"} qw|applied methods prerequisites|];
}

use Test::More;

my $obj_regex = qr/^Nuvol::([^:_]*)/;

# File

my @file_methods = qw|exists realpath|;
my @file_required
  = qw|_do_remove _do_slurp _do_spurt _from_file _from_host _from_url _get_download_url _to_host|;

sub test_file_applied ($file, $service) {
  my $role = "Nuvol::${service}::File";

  note 'File role';

  is $file->SERVICE, $service, "Service is $service";
  is $file->type, 'File', 'Type is File';
  ok $file->does($role), "Role $role applied";
  ok $file->does('Nuvol::Role::File'), 'File role applied';
}

sub test_file_prerequisites ($service) {
  my $role = "Nuvol::${service}::File";

  note 'Prerequisites for File';

  use_ok $role or BAIL_OUT "Unable to load $role";
  can_ok $role, $_ for @file_required;
}

sub test_file_methods ($file, $service) {
  ok $file->$_, "$_ returns a value" for @file_methods;
}

# Folder

my @folder_methods  = qw|exists realpath|;
my @folder_required = qw|_do_make_path _do_remove_tree|;

sub test_folder_applied ($folder, $service) {
  my $role = "Nuvol::${service}::Folder";

  note 'Folder role';

  is $folder->SERVICE, $service, "Service is $service";
  is $folder->type, 'Folder', 'Type is Folder';
  ok $folder->does($role), "Role $role applied";
  ok $folder->does('Nuvol::Role::Folder'), 'Folder role applied';
}

sub test_folder_prerequisites ($service) {
  my $role = "Nuvol::${service}::Folder";

  note 'Prerequisites for Folder';

  use_ok $role or BAIL_OUT "Unable to load $role";
  can_ok $role, $_ for @folder_required;
}

sub test_folder_methods ($folder, $service) {
  ok $folder->$_, "$_ returns a value" for @folder_methods;
}

# Metadata

my @metadata_methods  = qw|metadata description id name url|;
my @metadata_required = qw|_build_url _get_description _get_name _load_metadata|;

sub test_metadata_applied ($object, $service) {
  ref($object) =~ $obj_regex;
  my $role = "Nuvol::${service}::$1";

  note 'Metadata role';

  is $object->SERVICE, $service, "Service is $service";
  ok $object->does($role), "Role $role applied";

  ok $object->does('Nuvol::Role::Metadata'), "Metadata role applied";
  ok my $url = $object->url('abc99', 'def88'), 'Returns URL';
  isa_ok $url, 'Mojo::URL';
  is $url->path->[-1], 'def88', 'Path is extended';
}

sub test_metadata_methods ($object, $service) {
  note 'Metadata';
  ok $object->$_, "$_ returns a value" for grep !/id/, @metadata_methods;
}

sub test_metadata_prerequisites ($package, $service) {
  $package =~ $obj_regex;
  my $role = "Nuvol::${service}::$1";

  note 'Prerequisites for Metadata';

  use_ok $role or BAIL_OUT "Unable to load $role";
  can_ok $role, $_ for @metadata_required;
}

1;

=encoding utf8

=head1 NAME

Nuvol::Test::Roles - Test functions for Nuvol roles

=head1 SYNOPSIS

    # File
    use Nuvol::Test::Roles ':file';

    test_file_prerequisites;
    test_file_applied;
    test_file_methods;

    # Metadata
    use Nuvol::Test::Roles ':metadata';

    test_metadata_prerequisites;
    test_metadata_applied;
    test_metadata_methods;

=head1 DESCRIPTION

L<Nuvol::Test::Roles> provides test functions for Nuvol roles.

=head1 FUNCTIONS

=head2 test_file_applied

    test_file_applied $object, $service;

Tests if the L<File|Lib::Nuvol::Role::File> role and the service file role are applied.

=head2 test_file_methods

    test_file_methods $object, $service;    # live tests only

Tests if the L<File methods|Lib::Nuvol::Role::File/METHODS> return values.

=head2 test_file_prerequisites

    test_file_prerequisites $package, $service;

Tests if all the prerequisites for the L<File|Lib::Nuvol::Role::File> and the service file roles are met.

=head2 test_folder_applied

    test_folder_applied $object, $service;

Tests if the L<Folder|Lib::Nuvol::Role::Folder> role and the service folder role are applied.

=head2 test_folder_methods

    test_folder_methods $object, $service;    # live tests only

Tests if the L<Folder methods|Lib::Nuvol::Role::Folder/METHODS> return values.

=head2 test_folder_prerequisites

    test_folder_prerequisites $package, $service;

Tests if all the prerequisites for the L<Folder|Lib::Nuvol::Role::Folder> and the service folder roles are met.

=head2 test_metadata_applied

    test_metadata_applied $object, $service;

Tests if the L<Metadata|Lib::Nuvol::Role::Metadata> role is applied.

=head2 test_metadata_methods

    test_metadata_methods $object, $service;    # live tests only

Tests if the L<Metadata methods|Lib::Nuvol::Role::Metadata/METHODS> return values.

=head2 test_metadata_prerequisites

    test_metadata_prerequisites $package, $service;

Tests if all the prerequisites for the L<Metadata|Lib::Nuvol::Role::Metadata> role are met.

=head1 SEE ALSO

L<Nuvol::Test>, L<Nuvol::Test::Connector>, L<Nuvol::Test::Drive>, L<Nuvol::Test::Item>.

=cut


lib/Nuvol/Item.pm
package Nuvol::Item;
use Mojo::Base -base, -signatures;

use Mojo::File;
use Mojo::Path;

# constructor

sub new ($class, $drive, $params) {

  # check params
  Carp::croak 'Parameter metadata, id or path required!'
    unless $params->{metadata} || $params->{id} || $params->{path};
  Carp::croak q|Path must start with '/'| if $params->{path} && $params->{path} !~ m|^/|;

  my $self = bless {drive => $drive}, $class;

  my $service = $drive->SERVICE;
  $self->with_roles("Nuvol::${service}::Item");
  $self->{type} = $self->_get_type($params) or Carp::croak 'Unable to detect type!';

  $self->with_roles('Nuvol::Role::Metadata');
  $self->_parse_parameters($params);
  if ($self->{type} ne 'Unknown') {
    $self->with_roles("Nuvol::${service}::$self->{type}", "Nuvol::Role::$self->{type}");
  }

  return $self;
}

# methods

sub drive ($self)  { return $self->{drive}; }
sub exists ($self) { return $self->_check_existence; }
sub file ($self)   { return $self->{type} eq 'File'; }
sub folder ($self) { return $self->{type} eq 'Folder'; }

sub realpath ($self) {
  $self->{realpath} ||= Mojo::Path->new($self->_build_realpath)->leading_slash(1);
  return $self->{realpath};
}

sub type ($self) { return $self->{type}; }

# internal methods

# disables Nuvol::Role::Metadata::_load
sub _load ($self) {
  unless ($self->{metadata}) {
    $self->_set_metadata($self->_load_metadata);
    $self->{type} = $self->_get_type;
  }

  return $self;
}

1;

=encoding utf8

=head1 NAME

Nuvol::Item - Item in a drive

=head1 SYNOPSIS

    use Nuvol;
    my $drive = Nuvol::connect($configfile)->drive($drive_path);
    my $item  = $drive->item($path);

    $item->drive;
    $item->exists;
    $item->file;
    $item->folder;
    $item->realpath;
    $item->type;

    # files
    $item->copy_from;
    $item->copy_to;
    $item->spurt;
    $item->slurp;
    $item->download_url;
    $item->remove;

    # folders
    $item->make_path;
    $item->remove_tree;

    # metadata
    $item->description;
    $item->id;
    $item->metadata;
    $item->name;

=head1 DESCRIPTION

L<Nuvol::Item> is an item in a drive. It can be either a L<file|Nuvol::Role::File> or a
L<folder|Nuvol::Role::Folder>.

The syntax for drive items is oriented at L<Mojo::File>, so anyone familiar with this module will
recognize most of the methods.

=head1 CONSTRUCTOR

=head2 via Nuvol::Drive

    use Nuvol;
    $drive = Nuvol::Connector->new($configfile)->drive(%drive_path);

    $file   = $connector->item('/path/to/file');
    $folder = $connector->item('/path/to/folder/');

In daily use a L<Nuvol::Item> is created with L<Nuvol::Drive/item>. Paths must be absolute (starting
with a slash). Paths with trailing slash are interpreted as L<folders|Nuvol::Role::Folder>, without
slash as L<files|Nuvol::Role::File>.

=head2 new

    $item = Nuvol::Item->new($drive, {id       => $id});
    $item = Nuvol::Item->new($drive, {metadata => $metadata});
    $item = Nuvol::Item->new($drive, {path     => $path});

The internal constructor can be used if the C<id> or C<metadata> of the item are known.

=head1 METHODS

If a L<Nuvol::Item> is a file it inherits the following methods from L<Nuvol::Role::File>:

=over

=item L<copy_from|Nuvol::Role::File/copy_from>

=item L<copy_to|Nuvol::Role::File/copy_to>

=item L<spurt|Nuvol::Role::File/spurt>

=item L<slurp|Nuvol::Role::File/slurp>

=item L<download_url|Nuvol::Role::File/download_url>

=item L<remove|Nuvol::Role::File/remove>

=back

A folder inherits the following methods from L<Nuvol::Role::Folder>:

=over

=item L<make_path|Nuvol::Role::Folder/make_path>

=item L<remove_tree|Nuvol::Role::Folder/remove_tree>

=back

All item types inherit the following methods from L<Nuvol::Role::Metadata>:

=over

=item L<description|Nuvol::Role::Metadata/description>

=item L<id|Nuvol::Role::Metadata/id>

=item L<metadata|Nuvol::Role::Metadata/metadata>

=item L<name|Nuvol::Role::Metadata/name>

=item L<url|Nuvol::Role::Metadata/url>

=back

=head2 drive

    $drive = $item->drive;

Getter for the drive. Returns a L<Nuvol::Drive>.

=head2 exists

    $bool = $item->exists;

Checks if the item exists.

=head2 file

    $bool = $item->file;

Returns a true value if the L</type> of the item is C<File>.

=head2 folder

    $bool = $item->folder;

Returns a true value if the L</type> of the item is C<Folder>.

=head2 realpath

    $realpath = $item->realpath;

Getter for the full path of the item relative to its drive. Returns a L<Mojo::Path>.

    # '/path%20to/item'
    say $item->realpath;

    # '/path to/item'
    say $item->realpath->to_route;

Real paths are URL escaped. L<Mojo::Path/to_route> returns the unescaped string.

=head2 type

    $type = $item->type;

Getter for the type, can be C<File> or C<Folder>.

=head1 SEE ALSO

L<Mojo::File>, L<Nuvol::Drive>, L<Nuvol::Role::File>, L<Nuvol::Role::Folder>.

=cut


lib/Nuvol/Connector.pm
package Nuvol::Connector;
use Mojo::Base -base, -signatures;

use Mojo::Collection;
use Mojo::Parameters;
use Mojo::URL;
use Mojo::UserAgent;
use Nuvol::Config;
use Print::Colored ':all';

# constructor

sub new ($class, $configfile, $service = '', $params = {}) {
  my $self = bless {configfile => $configfile}, $class;

  # existing config
  if (-e $configfile) {
    my $config = $self->config;
    $self->with_roles("Nuvol::$config->{service}::Connector");
  }

  # create new config
  else {
    Carp::croak 'Service missing!' unless $service;
    $self->with_roles("Nuvol::${service}::Connector");
    my %config_params = (service => $self->SERVICE);
    for (sort keys $self->DEFAULTS->%*) {
      $config_params{$_} = $params->{$_} || $self->DEFAULTS->{$_}
        or Carp::croak "Parameter $_ missing!";
    }
    Nuvol::Config->new($configfile, \%config_params);
  }

  $self->with_roles('Nuvol::Role::Metadata');

  return $self;
}

# methods

sub authenticate ($self) {
  my $config = $self->config;

  # get authorization code
  my $response_type = $config->response_type;
  my %query         = (client_id => $config->app_id, response_type => $response_type);
  for (qw|scope redirect_uri|) {
    $query{$_} = $config->$_ if ($config->$_ || '') ne 'none';
  }
  my $url = Mojo::URL->new($self->AUTH_URL)->query(\%query);
  say $self->NAME;
  my $fn = $self->can("_auth_$response_type")
    or Carp::croak "Unsupported response type '$response_type'!";
  say qq|Open the following link in your browser and allow the application to access your drive:\n|;
  say_info qq|$url\n|;

  my $in       = prompt_input 'Paste the response URL: ', -stdio;
  my $response = $fn->($self, $config, Mojo::URL->new($in));

  # update config
  print 'Update config ... ';
  $self->authenticated($response);
  say_ok 'ok';

  return $self;
}

sub authenticated ($self, $params) {
  my $config = $self->config;

  $self->_set_token($config, $params);
  $config->save;

  return $self;
}

sub config ($self) {
  return Nuvol::Config->new($self->configfile);
}

sub configfile ($self) {
  return $self->{configfile};
}

sub disconnect ($self) {
  my $config = $self->config;
  $config->$_('') for qw|access_token refresh_token scope validto|;
  $config->save;

  return $self;
}

sub drive ($self, $path = '~') {
  require Nuvol::Drive;
  return Nuvol::Drive->new($self, {path => $path});
}

sub list_drives ($self) {
  require Nuvol::Drive;
  my $config = $self->config;

  my @drives   = $self->_load_drivelist;
  my $c_drives = Mojo::Collection->new(map { Nuvol::Drive->new($self, $_) } @drives);

  return $c_drives;
}

# internal methods

sub _access_token ($self) {
  my $config = $self->config;

  if ($config->validto && $config->validto < time + 10) {
    $self->_update_token($config);
    $config->save;
  }

  return $config->access_token;
}

sub _auth_code ($self, $config, $url) {
  my $code = $url->query->param('code') or die color_error 'URL contains no code!';

  # redeem code for access tokens
  print 'Get new token ... ';
  my $ua   = Mojo::UserAgent->new;
  my %form = (
    client_id  => $config->app_id,
    code       => $code,
    grant_type => 'authorization_code',
  );
  $form{redirect_uri} = $config->redirect_uri if $config->redirect_uri ne 'none';
  my $res = $ua->post($self->TOKEN_URL, form => \%form)->result;
  die color_error $res->message if $res->is_error;

  my $response = $res->json;
  say_ok 'ok';

  return $response;
}

sub _auth_headers ($self, $headers = {}) {
  my $access_token = $self->_access_token;
  $headers->{Authorization} = "Bearer $access_token";

  return $headers;
}

sub _auth_token ($self, $config, $url) {
  my $rv = {
    access_token =>Mojo::Parameters->new($url->fragment)->param('access_token'),
  };

  return $rv;
}

sub _set_token ($self, $config, $params) {
  $config->$_($params->{$_}) for qw|access_token refresh_token scope|;
  if ($params->{expires_in}) {
    $config->validto(time + $params->{expires_in});
  }

  return $self;
}

sub _ua_delete ($self, $url, @request) {
  return $self->_ua_tx(DELETE => $url, @request);
}

sub _ua_get ($self, $url, @request) {
  return $self->_ua_tx(GET => $url, @request);
}

sub _ua_post ($self, $url, @request) {
  return $self->_ua_tx(POST => $url, @request);
}

sub _ua_put ($self, $url, @request) {
  return $self->_ua_tx(PUT => $url, @request);
}

sub _ua_tx ($self, $method, $url, @request) {
  my $ua      = Mojo::UserAgent->new;
  my $headers = $self->_auth_headers(ref $request[0] eq 'HASH' ? shift @request : {});

  return $ua->start($ua->build_tx($method, $url, $headers, @request))->result;
}

sub _update_oauth_token ($self, $config) {
  my $ua   = Mojo::UserAgent->new;
  my $url  = $self->TOKEN_URL;
  my %form = (
    client_id     => $config->app_id,
    redirect_uri  => $config->redirect_uri,
    refresh_token => $config->refresh_token,
    grant_type    => 'refresh_token',
  );

  my $res = $ua->post($url, form => \%form)->result;
  Carp::confess $res->message if $res->is_error;

  $self->_set_token($config, $res->json);

  return $self;
}

1;

=encoding utf8

=head1 NAME

Nuvol::Connector - Base class for Nuvol connectors

=head1 SYNOPSIS

    # existing config file
    use Nuvol;
    my $configfile = '/path/to/configfile';
    my $connector  = Nuvol::connect($configfile);

    # new config file
    use Nuvol::Connector;
    my $connector = Nuvol::Connector->new($configfile, $service);

    # with additional parameters
    my $connector = Nuvol::Connector->new($configfile, $service, $params);

    $connector->authenticate;
    $connector->authenticated;
    $connector->config;
    $connector->configfile;
    $connector->disconnect;
    $connector->drive;
    $connector->drives;
    $connector->update_drives;
    $connector->url;

=head1 DESCRIPTION

L<Nuvol::Connector> is the base class for Nuvol connectors.

=head1 CONSTRUCTOR

=head2 via Nuvol

    use Nuvol;
    $configfile = '/path/to/configfile';
    $connector  = Nuvol::connect($configfile);

Connections with existing config files are opened with L<Nuvol::connect>.

=head2 new

    $connector = Nuvol::Connector->new($configfile, $service);

To create a new file, the parameter C<$service> is required. It defines the type of service that
will be activated in all objects created with this connector. Available services are
L<Dropbox|Nuvol::Connector::Dropbox>, L<Office365|Nuvol::Connector::Office365>, and
L<Dummy|Nuvol::Connector::Dummy>

    %params = (
      app_id       => $app_id,
      redirect_uri => $redirect_uri,
      scope        => $scope,
    );
    $connector = Nuvol::Connector->new($configfile, $service, \%params);

Optional parameters can be used to define C<app_id>, C<redirect_uri>, C<scope>. Notice that the final
value for the scope is set during authentication.

    $connector = Nuvol::Connector->new($configfile);

If the config file exists all values are read from it and additional parameters are ignored.
 
    $ perl -MNuvol::Connector -E'Nuvol::Connector->new("~/.office365.conf", "Office365")->authenticate'

New config files can be created and authenticated in the console.

=head1 METHODS

L<Nuvol::Connector> inherits the following methods from L<Nuvol::Role::Metadata>:

=over

=item L<description|Nuvol::Role::Metadata/description>

=item L<id|Nuvol::Role::Metadata/id>

=item L<metadata|Nuvol::Role::Metadata/metadata>

=item L<name|Nuvol::Role::Metadata/name>

=item L<url|Nuvol::Role::Metadata/url>

=back

=head2 authenticate

    $connector->authenticate;

Starts an interactive authentication process. It will display a URL that has to be opened in a
browser to start the authorization flow. The resulting URL is copied back to the console where it is
used to retrieve the access tokens. If this process succeeds, the information is written to the
config file. From now on it is possible to open the connector directly with L<Nuvol::connect>.

=head2 authenticated

    %params = (
      access_token  => $access_token,
      expires_in    => $seconds,
      refresh_token => $refresh_token,
      scope         => $scope,
    );
    $connector = $connector->authenticated(\%params);

Setter for new authentication tokens. Can be used if the authentication is made in an external
module.

=head2 config

    $value = $connector->config;

Getter for the config. Returns a L<Nuvol::Config>.

=head2 configfile

    $value = $connector->configfile;

Getter for the path to the configfile.

=head2 disconnect

    $connector = $connector->disconnect($newvalue);

Deletes the confidential information (access token, refresh token and valitidy time) from the config
file. On some services it invalidates the tokens on the server side, so it is always more secure to
use this method instead of just deleting the file.

To use a disconnected config file again it has to be re-authenticated with L</authenticate>.

=head2 drive

    $drive = $connector->drive($path);

Getter for a drive with the specified C<path>. Returns a L<Nuvol::Drive>.

    $drive = $connector->drive;
    $drive = $connector->drive('~');

Called with no parameter or with C<~> as path will return the default drive.

=head2 list_drives

    $drives        = $connector->list_drives;
    $default_drive = $connector->list_drives->first;

Lists the available drives. Returns a L<Mojo::Collection> containing one or more L<Nuvol::Drive>
with the default drive at the first position. This list may be incomplete.

=head1 SEE ALSO

L<Nuvol>, L<Nuvol::Config>, L<Nuvol::Drive>, L<Nuvol::Role::Metadata>.

=cut


lib/Nuvol.pm
package Nuvol;
use Mojo::Base -strict, -signatures;

our $VERSION = '0.02';

# functions

sub autoconnect ($configfile, $service) {
  require Nuvol::Connector;
  my $connector = Nuvol::Connector->new($configfile, $service);

  if (-f $configfile) {
    return $connector;
  } else {
    return $connector->authenticate;
  }
}

sub connect ($configfile) {
  require Nuvol::Connector;
  return Nuvol::Connector->new($configfile);
}

1;

=encoding utf8

=head1 NAME

Nuvol - A cloud toolkit

=head1 SYNOPSIS

    use Nuvol;

    # connect to a service
    my $configfile = '/path/to/configfile';
    my $service    = '...';  # one of Dropbox, Office365
    my $connector  = Nuvol::autoconnect($configfile, $service);

    # get main drive
    my $drive = $connector->drive('~');

    # upload a file
    use Mojo::File 'path';
    my $file = $drive->item('/My Text.txt')->copy_from(path 'Text on my PC.txt');

    # copy to another file
    my $file_2 = $file->copy_to('/path/to/Text Copy.txt');

    # download
    my $downloaded = $file_2->copy_to(path 'Downloaded Text.txt');

    # change the text
    $file_2->spurt('This text was changed.');

    # read it
    my $content = $file_2->slurp;

    # and delete the file
    $file_2->remove;

=head1 DESCRIPTION

L<Nuvol> is a toolkit to manipulate files and folders on cloud services. For
the beginning it supports L<Dropbox|Nuvol::Dropbox>, L<Office
365|Nuvol::Office365>, and a L<Dummy service|Nuvol::Dummy>.

    Nuvol
    └── Connector
        ├── Config
        │   └── config file
        └── Drive
            └── Item
                ├── File
                └── Folder

The services are organized in connectors, drives, items, files, and folders. The data needed to
access a service is stored in a config file.

=over 4

=item Connector

The L<Connector|Nuvol::Connector> is responsible for authentication and for the connection to the
cloud service.

=item Config and config file

The config file stores the tokens and other parameters used to establish a connection. Internally is
is represented by a L<Config|Nuvol::Config> object.

B<Warning:> The information in the config file allows full access to your cloud data for anyone who
can read it. It should be stored at a secure place. Services that are no longer used should be
disabled with L<Nuvol::Connector/disconnect>.

=item Drive

A L<Drive|Nuvol::Drive> is an isolated area where your data is stored. You may have a drive for your
personal and another for your business data. Not all cloud providers support different drives.

=item Item

Every object in a drive is an L<Item|Nuvol::Item>. Item is just an abstract type, a real object is
either a L<File|Nuvol::Role::File> or a L<Folder|Nuvol::Role::Folder>.

The syntax for drive items is oriented at L<Mojo::File>, so anyone familiar with this module will
recognize most of the methods.

=back

=head1 FUNCTIONS

None of the functions is exported.

=head2 autoconnect

    use Nuvol;
    $connector = Nuvol::autoconnect($configfile, $service);

Opens a connection using an existing config file, or starts an interactive
authentication process if the file doesn't exist. Returns a
L<Nuvol::Connector>.

=head2 connect

    use Nuvol;
    $connector = Nuvol::connect($configfile);

Opens a connection using an existing config file. Returns a L<Nuvol::Connector>.

=head1 AUTHOR & COPYRIGHT

© 2013–2020 by Tekki (Rolf Stöckli).

This program is free software, you can redistribute it and/or modify it under the terms of the
Artistic License version 2.0.

=head1 SEE ALSO

L<Nuvol::Connector>, L<Nuvol::Drive>, L<Nuvol::Item>, L<Nuvol::Test>.

=cut
